test/core/xds/xds_client_test.cc

```diff
@@ -1427,6 +1440,62 @@ TEST_F(XdsClientTest, ResourceValidationFailureForCachedResource) {
   }
 }
 
+TEST_F(XdsClientTest, WildcardCapableResponseWithEmptyResource) {
+  InitXdsClient();
+  // Start a watch for "wc1".
+  auto watcher = StartWildcardCapableWatch("wc1");
+  // Watcher should initially not see any resource reported.
+  EXPECT_FALSE(watcher->HasEvent());
+  // XdsClient should have created an ADS stream.
+  auto stream = WaitForAdsStream();
+  ASSERT_TRUE(stream != nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsWildcardCapableResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"wc1"});
+  CheckRequestNode(*request);  // Should be present on the first request.
+  // Server sends a response containing the requested resources plus an
+  // empty resource.
+  stream->SendMessageToClient(
+      ResponseBuilder(XdsWildcardCapableResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddWildcardCapableResource(XdsWildcardCapableResource("wc1", 6))
+          .AddEmptyResource()
+          .Serialize());
+  // XdsClient will delivery a valid resource update for wc1.
+  auto resource = watcher->WaitForNextResource();
+  ASSERT_TRUE(resource.has_value());
+  EXPECT_EQ(resource->name, "wc1");
+  EXPECT_EQ(resource->value, 6);
+  // XdsClient should NACK the update.
+  // There was one good resource, so the version will be updated.
+  request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsWildcardCapableResourceType::Get()->type_url(),
+               /*version_info=*/"1", /*response_nonce=*/"A",
+               // error_detail=
+               absl::InvalidArgumentError(absl::StrCat(
+                   "xDS response validation errors: ["
+                   "resource index 1: incorrect resource type \"\" "
+                   "(should be \"",
+                   XdsWildcardCapableResourceType::Get()->type_url(), "\")]")),
+               /*resource_names=*/{"wc1"});
+  // Cancel watch.
+  CancelWildcardCapableWatch(watcher.get(), "wc1");
+  // The XdsClient may or may not send an unsubscription message
+  // before it closes the transport, depending on callback timing.
```

Comment:
The way the code is currently written, it's not quite smart enough to realize at the moment that it unsubscribes that it doesn't actually need to do so if it happens to be removing the last subscription from that server.  All of the tests in this file do this same check.

Now that you mention it, though, it probably would be fairly easy to fix the code to be smarter about this.  I'll give that a try in a separate PR.

Reply:
It just looks like all the servers will just wait for the transport close since unsubscribe message is optional. This makes the message just noise... Until there will be that one server implementation that actually counts on unsubscribe message and has an intermittent memory leak 😉

Reply:
Yes, I agree that the client does not actually need to send the unsubscribe for the last resource if it is just going to close the stream immediately anyway; instead, it can close the stream without bothering to send the unsubscribe message first.  Ultimately, this is a question of optimization, not correctness, and it's not a case that anyone is actually likely to notice the sub-optimal behavior, but ideally we should be smarter about this.

I've put together #32166 to change the code to not bother sending the unsubscription in this case.